src/macros.rs

```diff
@@ -96,3 +96,232 @@ macro_rules! serde_string_impl_pk {
         }
     };
 }
+
+macro_rules! ImplDesc {
+    (@as_expr $e:expr) => { $e };
+
+    (
+        $trait_name:ident enum $struct_name:ident {
+            $(
+                $var_name:ident($var_ty:ident),
+            )*
+        }
+    ) => {
+        impl<Pk> $trait_name<Pk> for $struct_name<Pk> where Pk: MiniscriptKey {
+            /// See Trait level documentation for details.
+            fn sanity_check(&self) -> Result<(), Error> {
+                ImplDesc! {
+                    @as_expr
+                    match self {
+                        $(
+                            $struct_name::$var_name(ref v) => <$var_ty::<Pk> as $trait_name<Pk>>::sanity_check(&v),
+                        )*
+                    }
+                }
+            }
```

Comment:
It's hard to balance going too macro-crazy, but couldn't we make the user give every method name (and its signature) to the macro, and then generate all these from the same template?

This would make the macro smaller and also more general

Reply:
Yup, good idea. I think this should be possible